src/com/google/enterprise/gsafeed/GsafeedHelper.java

```diff
@@ -0,0 +1,125 @@
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+/**
+ * Helper methods for reading and creating GSA feed files.
+ */
+public class GsafeedHelper {
+
+  public static Gsafeed unmarshalWithDtd(URL url) throws JAXBException,
+      IOException, ParserConfigurationException, SAXException {
+    return unmarshalWithDtd(url.openStream());
+  }
+
+  /**
+   * Use the DTD to check for errors in the feed being read.
+   * Assume the DTD is on the classpath as /gsafeed.dtd.
+   */
+  public static Gsafeed unmarshalWithDtd(InputStream inputStream)
+      throws JAXBException, IOException, ParserConfigurationException,
+      SAXException {
+    JAXBContext jaxbContext =
+        JAXBContext.newInstance("com.google.enterprise.gsafeed");
+    Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
+    DocumentBuilderFactory docBuilderFactory =
+        DocumentBuilderFactory.newInstance();
+    docBuilderFactory.setValidating(true);
+    DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
+    docBuilder.setEntityResolver(new EntityResolver() {
+        @Override
+        public InputSource resolveEntity(String publicId,
+            String systemId) throws SAXException, IOException {
+          if (publicId.equals("-//Google//DTD GSA Feeds//EN")) {
+            return new InputSource(
+                GsafeedHelper.class.getResource("/gsafeed.dtd").openStream());
+          }
+          return null; // DTD is the only expected entity in feed files
+        }
+      });
+    docBuilder.setErrorHandler(new ErrorHandler() {
+        @Override
+        public void warning(SAXParseException exception)
+            throws SAXException {
+          throw new SAXException(exception.toString());
+        }
+
+        @Override
+        public void error(SAXParseException exception)
+            throws SAXException {
+          throw new SAXException(exception.toString());
+        }
+
+        @Override
+        public void fatalError(SAXParseException exception)
+            throws SAXException {
+          throw new SAXException(exception.toString());
+        }
+      });
+    Document document = docBuilder.parse(inputStream);
+    return (Gsafeed) unmarshaller.unmarshal(document);
+  }
+
+  /**
+   * Avoid reading the DTD. No validation will happen.
+   */
+  public static Gsafeed unmarshalWithoutDtd(URL url) throws JAXBException,
+      IOException, ParserConfigurationException, SAXException {
+    JAXBContext jaxbContext =
+        JAXBContext.newInstance("com.google.enterprise.gsafeed");
+    Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
+    DocumentBuilderFactory docBuilderFactory =
+        DocumentBuilderFactory.newInstance();
+    docBuilderFactory.setValidating(false);
+    DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
+    docBuilder.setEntityResolver(new EntityResolver() {
```

Comment:
Why no error handler here? Also, if we add one, is the only difference between these methods the argument to setValidating and the return value in resolveEntity?

Reply:
We're not reading the DTD and validating so it won't be checking for errors in the same way. Maybe we don't even want this option, but it's easier to get suggestions when code exists so I put it in. Or possibly we don't want to validate feed files.